### ISSUE TITLE
model.wv.accuracy messes up the model object

### ISSUE THREAD

**Issue by ahmedahmedov** [RaRe-Technologies/gensim#2014]
I was playing around with model.wv.accuracy to test the accuracy of my model. It worked fine when the only parameter I set was the filename:

`model.wv.accuracy("questions-words.txt")
`
Things get weird though when I set the 'most_similar' parameter of the function: 

`model.wv.accuracy("questions-words.txt", most_similar=gensim.models.KeyedVectors.most_similar_cosmul)`

First, it gives an error saying that gensim.models.KeyedVectors.most_similar_cosmul doesn't have a parameter named 'retrict_vocabulary'. 

Most importantly, right after that, the model vocabulary is messed up for some reason. That is, when calling model.wv.accuracy("question-words.txt") again, I get the following error: 

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-36-41dd7cacf846> in <module>()
----> 1 model.wv.accuracy("questions-words.txt", most_similar=gensim.models.KeyedVectors.most_similar_cosmul)
      2 model.wv.evaluate_word_pairs("wordsim353.tsv")

/home/user/anaconda3/lib/python3.5/site-packages/gensim/models/keyedvectors.py in accuracy(self, questions, restrict_vocab, most_similar, case_insensitive)
    772 
    773         """
--> 774         ok_vocab = [(w, self.vocab[w]) for w in self.index2word[:restrict_vocab]]
    775         ok_vocab = {w.upper(): v for w, v in reversed(ok_vocab)} if case_insensitive else dict(ok_vocab)
    776 

/home/user/anaconda3/lib/python3.5/site-packages/gensim/models/keyedvectors.py in <listcomp>(.0)
    772 
    773         """
--> 774         ok_vocab = [(w, self.vocab[w]) for w in self.index2word[:restrict_vocab]]
    775         ok_vocab = {w.upper(): v for w, v in reversed(ok_vocab)} if case_insensitive else dict(ok_vocab)
    776 

KeyError: 'the'
```
calling model.wv.most_similar also gives the same error.

**Comment by ahmedahmedov** [comment]
I think I know why it happened. For some reason, when setting case_insensitive to True, the vocabulary is converted to upper case. model.accuracy should not make any changes to the internal state of the model.

**Comment by piskvorky** [comment]
Definitely a bug. I'd also expect this function to be 100% "const" (not modifying the object). Thanks for reporting!

**Comment by ahmedahmedov** [comment]
Can someone please explain me what is going in the following piece of code extracted from gensim.models.keyedvectors.py?

```
ignore = {a, b, c}  # input words to be ignored
predicted = None
# find the most likely prediction, ignoring OOV words and input words
sims = most_similar(self, positive=[b, c], negative=[a], topn=False, restrict_vocab=restrict_vocab)
self.vocab = original_vocab
for index in matutils.argsort(sims, reverse=True):
    predicted = self.index2word[index].upper() if case_insensitive else self.index2word[index]
    if predicted in ok_vocab and predicted not in ignore:
        if predicted != expected:
            logger.debug("%s: expected %s, predicted %s", line.strip(), expected, predicted)
        break
if predicted == expected:
    section['correct'].append((a, b, c, expected))
else:
    section['incorrect'].append((a, b, c, expected))
```
What is the logic behind this approach? 

1. given the analogy,` a->b as c->d`, first, `sims = most_similar(self, positive=[b, c], negative=[a], topn=False, restrict_vocab=restrict_vocab` is obtained. The sims object contains the list of distances of the vector (b+c-a) to all the other vectors. Why?
2. The sims list is sorted in reverse, obtaining the indices of the words in the reverse order.
3. For each index in the vocabulary, its corresponding word is obtained and simply used for logging purposes!
4. after the for loop exits, predicted ends up assigned to the last word in the list
5. if predicted==expected, then add it to section['correct'] if not then to section['incorrect']

This is totally incorrect and I believe the results obtained with this code is invalid. I think it is an urgent case. The code needs to be reversed to the last working version.

**Comment by menshikh-iv** [comment]
Hello @ahmedahmedov

> First, it gives an error saying that gensim.models.KeyedVectors.most_similar_cosmul doesn't have a parameter named 'retrict_vocabulary'.

yes, the signature of `most_similar` function should be like https://github.com/RaRe-Technologies/gensim/blob/5677ab300e4e3dc4645806762b693902b97c13c3/gensim/models/keyedvectors.py#L363

more accurately (as this used in `accuracy`)
```python
sims = most_similar(self, positive=[b, c], negative=[a], topn=False, restrict_vocab=restrict_vocab)
```
`most_similar_cosmul` doesn't match to this signature (probably should be, looks like a bug for me).
I also agree with Radim (about internal state), can you make a PR with fix @ahmedahmedov (this will be really great)?

**Comment by ahmedahmedov** [comment]
I would like to suggest to replace https://github.com/RaRe-Technologies/gensim/blob/5677ab300e4e3dc4645806762b693902b97c13c3/gensim/models/keyedvectors.py#L979-L992

with 

```
predicted = None
# find the most likely prediction, ignoring OOV words and input words
predicted_tuple = model.most_similar(positive=[b, c], negative=[a], topn=1, restrict_vocab=restrict_vocab)
predicted = predicted_tupe[0].upper() if case_insensitive else predicted_tuple[0]
model.wv.vocab = original_vocab

if predicted not in ignore and predicted==expected:
    section['correct'].append((a, b, c, word[0]))

else:
    section['incorrect'].append((a, b, c, expected))
```

The idea is, there is no need to calculate similarity stuff separately (unless there is an important reason behind this design choice). It is already done inside the `most_similar` method. Just set `topn=1` and see if the result matches the expected value. If yes, then update `section['correct']`, if not update `section['incorrec't]`. In addition, we can set `topn` as a parameter to accuracy to give users the flexibility to decide what they consider as accurate.

**Comment by ahmedahmedov** [comment]
it seems that case_insensitive dictionary search is not that easy in Python. Any ideas how we might find a workaround to modify the most_similar method to be able to search in a case_insensitive manner without having to create an additional vocabulary? 
[Python dictionary case_insensitive search](https://stackoverflow.com/questions/3296499/case-insensitive-dictionary-search/3296782#3296782)

**Comment by piskvorky** [comment]
How does your suggestion handle the `ignoring OOV words and input words`?

**Comment by ahmedahmedov** [comment]
`if predicted not in ignore and predicted==expected` handles input words. If the word is not found in the vocabulary, we can simply skip it, right? Then, surrounding most_similar with` try-excep`t clause and ignoring when `KeyIsNotFound` exception is raised should solve the OOV case.

**Comment by piskvorky** [comment]
No, that would change the logic in a fundamental way. We want to keep the "standard" logic from the C tool, which looks for the nearest word that is not input/OOV.

**Comment by piskvorky** [comment]
I agree with your analysis.

Re. 4: in case of collisions, the original C tool simply keeps the most frequent candidate. Like you say, doing this without vocabulary copies (lots of RAM) is much preferable.

I also saw another technique to reduce memory / collisions, by collapsing similar words together (when vectors for `Apple` and `APPLE` are very similar, just keep one). Only tangentially related here, mentioning just for completeness :)

**Comment by ahmedahmedov** [comment]
> I'd say no, that would change the logic in a fundamental way. We want to keep the "standard" logic from the C tool, which looks for the nearest word that is not input/OOV.

https://github.com/RaRe-Technologies/gensim/blob/5677ab300e4e3dc4645806762b693902b97c13c3/gensim/models/keyedvectors.py#L979-L992

as far as I can see, OOV words are handled within model.wv.word_vec which throws KeyNotFound exception when a word is not found in the vocabulary:

https://github.com/RaRe-Technologies/gensim/blob/49e6abdde6bca55b7a5b263c01ecdc94a75fb1ab/gensim/models/keyedvectors.py#L326-L335

Again, I am sorry if I am missing something very obvious here, or I have failed to explain what I really want to achieve. (I will update the comment later to write a step-by-step explanation)

**Comment by piskvorky** [comment]
Here, OOV = words that are not a part of the subset of words we care about (because of vocabulary size restrictions; IIRC the C tool considers only the top 30k most frequent words and their vectors).

So, what happens in that loop is that `accuracy` searches for the first word (=highest similarity) that is **both** not OOV and not any of the 3 input words. Once it finds such word, it checks whether it's equal to the 4th word (case-insensitive, IIRC).

**Comment by Akashtyagi08** [comment]
Does this bug still exists ? When tried using model.wv.accuracy() it shows deprecation warning now.
